### PR TITLE
Haydn: Make SoundIO session implementation lock-free

### LINKED ISSUE #2062: SoundIO Object Reference not set.
### What's the issue you encountered?

I was playing Persona 5 Strikers, and after getting Sophia to join my party, the game immediately crashed as Joker started dashing. A soundIO error</br>
Fresh ryujinx install! </br>

### How can the issue be reproduced?

Play Persona 5 Strikers, then play until you get to Sophia, after moving from the battle, you should get the error. (Make sure to use SoundIO backend!)

### Log file
[Ryujinx_1.0.6732_2021-02-26_10-27-55.log](https://github.com/Ryujinx/Ryujinx/files/6050746/Ryujinx_1.0.6732_2021-02-26_10-27-55.log)
```
Application : Unhandled exception caught: System.NullReferenceException: Object reference not set to an instance of an object.
   at Ryujinx.Audio.Backends.SoundIo.SoundIoHardwareDeviceSession.Update(Int32 minFrameCount, Int32 maxFrameCount) in C:\projects\ryujinx\Ryujinx.Audio.Backends.SoundIo\SoundIoHardwareDeviceSession.cs:line 414

```
### Environment?

- Ryujinx version: 1.0.6732</br>
- Game version: 1.0.0</br>
- System Specs:
  - OS: *Windows 10 Professional Build 19042*
  - CPU: *Ryzen 9 3900X*
  - GPU: *5700XT*
  - RAM: *32GB*
- Applied Mods: None, I just have the DLC Legacy BGM & All-Out Attack Pack DLC.

### PR DESCRIPTION
**Pull request by Thog**
This should fix weird crashes happening for some people with SoundIO.

It also is expected to improve performances a bit.

Fix #2062

### CODE REVIEW

File: Ryujinx.Audio.Backends.SoundIo/SoundIoHardwareDeviceSession.cs [hunk 1]
```diff
@@ -411,22 +411,25 @@ private unsafe void Update(int minFrameCount, int maxFrameCount)
 
             bool needUpdate = false;
 
-            while (availaibleSampleCount > 0 && _queuedBuffers.TryPeek(out SoundIoAudioBuffer driverBuffer))
+            lock (_lock)
             {
-                ulong sampleStillNeeded = driverBuffer.SampleCount - driverBuffer.SamplePlayed;
-                ulong playedAudioBufferSampleCount = Math.Min(sampleStillNeeded, availaibleSampleCount);
+                while (availaibleSampleCount > 0 && _queuedBuffers.TryPeek(out SoundIoAudioBuffer driverBuffer))
```
Comment (Shahil-Ayato) [inline Ryujinx.Audio.Backends.SoundIo/SoundIoHardwareDeviceSession.cs:416]: Shouldn’t this be Available Instead of availaible?

Comment (Thog) [reply Ryujinx.Audio.Backends.SoundIo/SoundIoHardwareDeviceSession.cs:416]: No it's a local variable

Comment (gdkchan) [reply Ryujinx.Audio.Backends.SoundIo/SoundIoHardwareDeviceSession.cs:416]: This was not resolved, the issue is the typo on the variable name.

Comment (Shahil-Ayato) [reply Ryujinx.Audio.Backends.SoundIo/SoundIoHardwareDeviceSession.cs:416]: Yeah I didn’t mean to capitalize the A, so my mistake 🥴

**Comment by gdkchan** [comment]
Note that .NET has `ConcurrentQueue` that is thread safe, if possible, that should be preferred over using a `Queue` with a lock, as it's likely to have better performance.

File: Ryujinx.Audio.Backends.SoundIo/SoundIoHardwareDeviceSession.cs (hunk 1 above)
Comment (Thog) [reply Ryujinx.Audio.Backends.SoundIo/SoundIoHardwareDeviceSession.cs:416]: aaah that *not like that*

**Comment by Thog** [comment]
> 
> 
> Note that .NET has `ConcurrentQueue` that is thread safe, if possible, that should be preferred over using a `Queue` with a lock, as it's likely to have better performance.

I think I'm going for it, will turn the ``_playedSampleCount`` into an atomic and update the OpenAL codepaths

File: Ryujinx.Audio.Backends.SoundIo/SoundIoHardwareDeviceSession.cs (hunk 1 above)
Comment (Thog) [reply Ryujinx.Audio.Backends.SoundIo/SoundIoHardwareDeviceSession.cs:416]: Will fix in a separated PR.

**Review by AcK77 — APPROVED** [review]
Lgtm!

**Review by riperiperi — APPROVED** [review]
Nice one.

**Review by gdkchan — APPROVED** [review]
lgtm, thanks. Nice to see a lock-free implementation.

**Pull request merged**